### PR TITLE
refactor conversion tests

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
refactor conversion tests to cover more edge cases. also fix some minor bugs in literal handling, and update the bounds for decimals in generated code.

### How I did it
added more edge cases, and rewrote the test cases around a core "spec" function, `_py_convert`, which implements the conversion rules on python objects.

### How to verify it
:eyes: 

### Commit message

Commit message for the final, squashed PR. (Optional, but reviewers will appreciate it! Please see [our commit message style guide](../../blob/master/docs/style-guide.rst#best-practices-1) for what we would ideally like to see in a commit message.)

### Cute Animal Picture

image

### CODE REVIEW

**Review by fubuloubu — APPROVED** [review]

File: tests/parser/functions/test_convert.py [hunk 1]
```diff
@@ -1,544 +1,532 @@
...
-            return ot_bytes_info.m_bits >= ADDRESS_BITS
-        return it in ["decimal", "Bytes", "address"]
+            return i_detail.type_bytes <= o_detail.type_bytes
+
+        # the fact that any bytes can be converted to decimal but not all ints
+        # may be an inconsistency in the spec
+        return i_detail.type_class in ("decimal", "bytes")
+
+    elif i_detail.type_class == "Bytes":
+        return o_detail.type_class in ("int", "decimal", "address")
```
Comment (tserg) [inline tests/parser/functions/test_convert.py:126]: Do we want to test conversion of `Bytes[N]` to `bytesM` where M >= N?

File: tests/parser/functions/test_convert.py [hunk 2]
```diff
@@ -1,544 +1,532 @@
...
+    elif typename == "decimal":
+        info = parse_decimal_info(typename)
+        assert info.bits % 8 == 0
+        return TestType(typename, info.bits // 8, "decimal", info)
+    elif typename.startswith("bytes"):
+        info = parse_bytes_m_info(typename)
+        return TestType(typename, info.m, "bytes", info)
+    elif typename.startswith("Bytes"):
+        assert typename == "Bytes[32]"  # TODO test others
+        return TestType(typename, 32, "Bytes", info)
```
Comment (tserg) [inline tests/parser/functions/test_convert.py:80]:
```suggestion
        return TestType(typename, 32, "Bytes", None)
```

File: tests/parser/functions/test_convert.py [hunk 3]
```diff
@@ -1,544 +1,532 @@
...
+    elif typename.startswith("String"):
+        assert typename == "String[32]"  # TODO test others
+        return TestType(typename, 32, "Bytes", info)
```
Comment (tserg) [inline tests/parser/functions/test_convert.py:83]:
```suggestion
        return TestType(typename, 32, "Bytes", None)
```

File: tests/parser/functions/test_convert.py [hunk 4]
```diff
@@ -1,544 +1,532 @@
...
+        return i_detail.type_class in ("decimal", "bytes")
+
+    elif i_detail.type_class == "Bytes":
+        return o_detail.type_class in ("int", "decimal", "address")
+
+    elif i_typ == "decimal":
+        if o_detail.type_class == "bytes":
+            return i_detail.type_bytes <= o_detail.type_bytes
+
+        return o_detail.type_class in ("int", "bool")
```
Comment (tserg) [inline tests/parser/functions/test_convert.py:132]: Do we want to test conversion of `decimal` to `Bytes[N]`?

File: tests/parser/functions/test_convert.py [hunk 5]
```diff
@@ -1,544 +1,532 @@
...
+
+    elif i_typ == "address":
+        return o_typ in ("uint", "bytes")
```
Comment (tserg) [inline tests/parser/functions/test_convert.py:135]: Do we want to test conversion of `address` to `Bytes[N]`?

Comment (tserg) [inline tests/parser/functions/test_convert.py:135]:
```suggestion
        if o_detail.type_class == "bytes":
            return i_detail.type_bytes < o_detail.type_bytes
        elif o_detail.type_class == "int":
            return not o_detail.info.is_signed
        return False
```
We need to filter for `bytesN` where `N` > 20, and for `int` in the case of "int" type_class.

**Review by tserg — CHANGES_REQUESTED** [review]

File: tests/parser/functions/test_convert.py (hunk 5 above)
Comment (charles-cooper) [reply tests/parser/functions/test_convert.py:135]: it's not implemented yet :sweat_smile: see https://github.com/vyperlang/vyper/blob/7e0cf2ad8c6bbbbca4b1f6dd77c69b20d81393ca/vyper/builtin_functions/convert.py#L386

Comment (charles-cooper) [reply tests/parser/functions/test_convert.py:135]: why is that?

File: tests/parser/functions/test_convert.py (hunk 4 above)
Comment (charles-cooper) [reply tests/parser/functions/test_convert.py:132]: also not implemented yet

File: tests/parser/functions/test_convert.py (hunk 1 above)
Comment (charles-cooper) [reply tests/parser/functions/test_convert.py:124]: oh, never mind, i see. these could probably be tested.
